### PR TITLE
Adding docs for docstrings and typehints

### PR DESCRIPTION
**Pull request by seankane-msft**

### CODE REVIEW

File: doc/dev/docstring_typehint.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+Docstrings are noted by the Python long-string `"""<docstring>"""`. When adding docstrings the `~` can be optionally added to the front of a custom model, this will only display the actual model name in the documentation. For example a type that looks like `azure.data.tables.TableEntity` will display the entire path in the documentation, but `~azure.data.tables.TableEntity` will only display `TableEntity` in the documentation. Additionally, adding the wrapper `` :class:`~azure.data.tables.TableEntity` `` will display only `TableEntity` formatted in monospace font. Both of these capabilities are optional and can only be used in the two-line format described below.
+
+### Method Docstrings
+
+A method docstring is annotated by the Python long-string `"""<docstring>"""` right after the method definition. Below is an example of a full docstring from `azure-ai-formrecognizer`:
+```python
+    @distributed_trace
+    def begin_training(self, training_files_url, use_training_labels, **kwargs):
+        # type: (str, bool, Any) -> LROPoller[CustomFormModel]
+        """Create and train a custom model. The request must include a `training_files_url` parameter that is an
```
Comment (lmazuel) [inline doc/dev/docstring_typehint.md:23]: Convention is a short line finished by a dot and two new lines:
```
"""I am a short definition.

The rest of it in long text
```

File: doc/dev/docstring_typehint.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        :param bool use_training_labels: Whether to train with labels or not. Corresponding labeled files must
+            exist in the blob container if set to `True`.
+        :keyword str prefix: A case-sensitive prefix string to filter documents in the source path for
+            training. For example, when using a Azure storage blob URI, use the prefix to restrict sub
+            folders for training.
+        :keyword bool include_subfolders: A flag to indicate if subfolders within the set of prefix folders
+            will also need to be included when searching for content to be preprocessed. Not supported if
+            training with labels.
+        :keyword str model_name: An optional, user-defined name to associate with your model.
+        :keyword str continuation_token: A continuation token to restart a poller from a saved state.
```
Comment (lmazuel) [inline doc/dev/docstring_typehint.md:41]: We should have an example with inline type and clarify that for `keyword` you then need to use `paramtype`, while a regular `param` should be `type`

Comment (lmazuel) [reply doc/dev/docstring_typehint.md:41]: You say that later, but still this overview should show it

File: doc/dev/docstring_typehint.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                :language: python
+                :dedent: 8
+                :caption: Training a model (without labels) with your custom forms.
+        """
+        ...
+```
+
+The first portion of this docstring is a general description of what the method does. Following the general description of the method is a **required new line** and then documentation for each of the parameters, optional keyword arguments, returned objects, and potentially raised errors.
+
+Required parameters can be documented in one-line or two-lines. Both options can be used in a single docstring for documenting different parameters without issue.
```
Comment (lmazuel) [inline doc/dev/docstring_typehint.md:66]: Not "required", but positional:

```
def (a,b,c=None)
```
They are all positional and documented with `param`, but they are not all required.

File: doc/dev/docstring_typehint.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+```python
+:param <type> <param_name>: <Description of the parameter>
+```
+2. This option works best for parameters that use custom models or have many different options
+```python
+:param <param_name>: <Description of the parameter>
+:type <param_name>: <param_type>
+```
+
+Optional keyword arguments can in one-line or two-lines. Both options can be used in a single docstring for documenting different keywords without issue.
```
Comment (lmazuel) [inline doc/dev/docstring_typehint.md:77]: keyword includes kwargs and parameters after the * character if you're on Python 3 code

File: doc/dev/docstring_typehint.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+from typing import List, Dict, Union
+from ._models import TextDocumentInput, RecognizeEntitiesResult
+
+def recognize_entities(
+    self,
+    documents: Union[List[str], List[TextDocumentInput], List[Dict[str, str]]],
+    **kwargs: Any
+) -> List[Union[RecognizeEntitiesResult, DocumentError]]:
+```
+Note that in the in-line type hints the custom models do not to be guarded by an `if TYPE_CHECKING` conditional. These custom models must be included in imports or the program will fail on the type not being found.
```
Comment (lmazuel) [inline doc/dev/docstring_typehint.md:185]: You probably want a note to say "Dont' use `dict[str,[list[str]]]` the lower case version of type"

File: doc/dev/docstring_typehint.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+:rtype: <Type of the returned object>
+"""
+```
+
+Finally, describe the possible errors raised by a method:
+```python
+:raises <error1>, <error2>, or <error3>: <description>
+```
+
+All of the above information needs to be added for each public method and there **needs to be a newline after the above docstrings**. There are additional options for including examples and version specific additions.
```
Comment (kristapratico) [inline doc/dev/docstring_typehint.md:101]: for additional options, not sure if you want to link out to this: https://review.docs.microsoft.com/en-us/help/onboard/admin/reference/python/documenting-api?branch=master#supported-sphinx-directives

File: doc/dev/docstring_typehint.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+For adding a version specific change use the `versionadded` docstring:
+```
+.. versionadded:: <version_number>
+
+```
+
+### Model and Client Docstrings
+
+For documenting the properties of a model or client, include a docstring right after the `__init__` method or immediately after the class declaration
```
Comment (kristapratico) [inline doc/dev/docstring_typehint.md:111]: Models that are used in requests need to have more than just the ivars documented. Wondering if we should include an example for that as well.

File: doc/dev/docstring_typehint.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+## Type Hints
+
+### Type Hints for Python 2.7 and 3.5+
+
+Python 2.7 does not support in-line type hints, the type hints for all sync code will have to adhere to type-hints in comments only. On all public methods include a type-hint by using the `# type: (...) -> (...)` format:
+```python
+def add(num1, num2):
+    # type: (int, int) -> int
+    ...
```
Comment (kristapratico) [inline doc/dev/docstring_typehint.md:145]: maybe show in the snippet that `add()` is returning an `int` so the return type comment makes sense

File: doc/dev/docstring_typehint.md (hunk 2 above)
Comment (seankane-msft) [reply doc/dev/docstring_typehint.md:41]: Half use inline type, half use paramtype.

File: doc/dev/docstring_typehint.md (hunk 7 above)
Comment (seankane-msft) [reply doc/dev/docstring_typehint.md:111]: Could you provide a hint for me on that and I can include it?

File: doc/dev/docstring_typehint.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            https://docs.microsoft.com/azure/cognitive-services/form-recognizer/build-training-data-set
+        :param bool use_training_labels: Whether to train with labels or not. Corresponding labeled files must
+            exist in the blob container if set to `True`.
+        :keyword str prefix: A case-sensitive prefix string to filter documents in the source path for
+            training. For example, when using a Azure storage blob URI, use the prefix to restrict sub
+            folders for training.
+        :keyword include_subfolders: A flag to indicate if subfolders within the set of prefix folders
+            will also need to be included when searching for content to be preprocessed. Not supported if
+            training with labels.
+        :paramtype include_subfolder: bool
```
Comment (kristapratico) [inline doc/dev/docstring_typehint.md:42]: nit: we lost the plurality of the kwarg name

```suggestion
        :paramtype include_subfolders: bool
```

File: doc/dev/docstring_typehint.md (hunk 7 above)
Comment (kristapratico) [reply doc/dev/docstring_typehint.md:117]: Sure, the input for document translation here: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/translation/azure-ai-translation-document/azure/ai/translation/document/_models.py#L150

File: doc/dev/docstring_typehint.md (hunk 9 above)
Comment (kristapratico) [inline doc/dev/docstring_typehint.md:42]: nit: we lost the plurality of the kwarg name

```suggestion
        :paramtype include_subfolders: bool
```

File: doc/dev/README.md [hunk 10]
```diff
@@ -1,10 +1,16 @@
-# Dev documentation
+# Developer Documentation
 
 This folder contains documentation for developers of SDK: internal teams at Microsoft, or advanced contributors.
```
Comment (mccoyp) [inline doc/dev/README.md:3]:
```suggestion
This folder contains documentation for developers of SDKs: internal teams at Microsoft, or advanced contributors.
```

**Review by mccoyp — COMMENTED** [review]
As usual, just a nitpicking comment from me. Looks good!

**Review by kristapratico — APPROVED** [review]

**Pull request merged**